core/src/main/scala/kafka/server/BrokerServer.scala

```diff
@@ -499,12 +499,13 @@ class BrokerServer(
       if (clientToControllerChannelManager != null)
         CoreUtils.swallow(clientToControllerChannelManager.shutdown(), this)
 
-      if (logManager != null)
-        CoreUtils.swallow(logManager.shutdown(), this)
-      // be sure to shutdown scheduler after log manager
+      // be sure to shutdown scheduler before log manager
```

Comment:
(1) Could we add some comment to explain why we want to shut down the log manager later?
(2) If we are shutting down, it makes sense to stop the scheduler first since there is no guarantee any asynchronously scheduled task will complete. If this is the case, should we shut down the scheduler before any other component (e.g. ReplicaManger also uses scheduler)?
(3) Once the scheduler is shut down, scheduling new tasks causes IllegalStateException. That's probably the exception that https://github.com/apache/kafka/pull/10538 tries to fix. To avoid polluting the log, perhaps we could change KafkaScheduler such that it avoids throwing an exception once it's shutdown.

Reply:
Are all tasks scheduled of the async variety? It would be nice if we would stop issuing new async tasks when shutdown starts, but let the existing ones complete.

Reply:
It seems that none of the scheduled calls depends on the completion of the returned future. When shutting down KafkaScheduler, we call ScheduledThreadPoolExecutor.shutdown(), which doesn't wait for all existing tasks to complete. This seems ok since for important tasks (e.g. flushing the log), we make explicit calls during shutdown.

Reply:
Thanks for the code review. I moved `kafkaScheduler.shutdown` upfront, changed the exception to an info-level log, and added comments.

For the exception to log change: I looked at the code that calls `startup` and `shutdown`. It appears we always call startup right after creating the scheduler object and call shutdown when calling the parent's shutdown or close. I think the callers make sure that they won't call `scheduler.schedule` to a scheduler that has been shutdown, unless we are shutting down the broker. So it should be okay to change the `IllegalStateException` to an info-level log

Reply:
@junrao `shutdown`:

"Initiates an orderly shutdown in which previously submitted tasks are executed, but no new tasks will be accepted. Invocation has no additional effect if already shut down."

Maybe you were thinking of `shutdownNow`?

Reply:
IIUC, the code (1) lets running tasks to complete, (2) waits for tasks scheduled with `executor.schedule(runnable, delay, unit)` to finish after the delay, and (3) cancels future tasks scheduled by `executor.scheduleAtFixedRate(runnable, delay, period, unit)`

https://github.com/apache/kafka/blob/b61ec0003f907b61243102811fdb2e92f8d7d2c5/core/src/main/scala/kafka/utils/KafkaScheduler.scala#L121-L124

Reply:
@ijuma and @ccding :  The description of ScheduledThreadPoolExecutor.shutdown() says 

`This method does not wait for previously submitted tasks to complete execution. Use awaitTermination to do that.
`

So, we probably want to call awaitTermination() with a timeout like 10secs to make sure all existing tasks complete before shutting down other components.

We probably can't use shutdownNow() since it interrupts the task and could cause IOException when blocking operations (e.g., force) are applied on a file channel. This will then lead to unclean shutdown.

Reply:
Yeah, it does not wait for previously submitted tasks to complete execution, also does not kill the tasks. KafkaScheduler does call `awaitTermination` with a 1-day timeout. Do you think we want a shorter timeout?

Reply:
@ccding : Missed that part. `awaitTermination` with a 1-day timeout is fine.